When svg files are not loadable, revert to png file for throttles icons.

Lionel642 opened the pull request:
For [issue 9761](https://github.com/JMRI/JMRI/issues/9761)

dheap commented:
@Lionel642  Thanks.

dheap commented:
Unfortunately still fails:
```
021-05-09 08:06:23,962 swing.ResizableImagePanel             ERROR - resources/icons/functionicons/svg/lightsOff.svg is not a valid image file, exception:  [AWT-EventQueue-0]
net.coobird.thumbnailator.tasks.UnsupportedFormatException: No suitable ImageReader found for resources/icons/functionicons/svg/lightsOff.svg.
	at net.coobird.thumbnailator.tasks.io.FileImageSource.read(Unknown Source)
	at jmri.util.swing.ResizableImagePanel.readImage(ResizableImagePanel.java:165)
	at jmri.util.swing.ResizableImagePanel.setImagePath(ResizableImagePanel.java:216)
	at jmri.jmrit.throttle.FunctionButton.setIconPath(FunctionButton.java:512)
	at jmri.jmrit.throttle.FunctionPanel.resetFnButtons(FunctionPanel.java:262)
	at jmri.jmrit.throttle.FunctionPanel.initGUI(FunctionPanel.java:168)
	at jmri.jmrit.throttle.FunctionPanel.<init>(FunctionPanel.java:45)
	at jmri.jmrit.throttle.ThrottleFrame.initGUI(ThrottleFrame.java:285)
	at jmri.jmrit.throttle.ThrottleFrame.<init>(ThrottleFrame.java:99)
	at jmri.jmrit.throttle.ThrottleWindow.initGUI(ThrottleWindow.java:101)
	at jmri.jmrit.throttle.ThrottleWindow.<init>(ThrottleWindow.java:87)
	at jmri.jmrit.throttle.ThrottleFrameManager.createThrottleWindow(ThrottleFrameManager.java:44)
	at jmri.jmrit.throttle.ThrottleFrameManager.createThrottleFrame(ThrottleFrameManager.java:59)
	at jmri.jmrit.throttle.ThrottleCreationAction.actionPerformed(ThrottleCreationAction.java:62)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.AbstractButton.doClick(AbstractButton.java:376)
	at com.apple.laf.ScreenMenuItem.actionPerformed(ScreenMenuItem.java:125)
	at java.awt.MenuItem.processActionEvent(MenuItem.java:669)
	at java.awt.MenuItem.processEvent(MenuItem.java:628)
	at java.awt.MenuComponent.dispatchEventImpl(MenuComponent.java:357)
	at java.awt.MenuComponent.dispatchEvent(MenuComponent.java:345)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:763)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:84)
	at java.awt.EventQueue$4.run(EventQueue.java:733)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:730)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
2021-05-09 08:06:23,971 swing.ResizableImagePanel             ERROR - resources/icons/functionicons/svg/lightsOn.svg is not a valid image file, exception:  [AWT-EventQueue-0]
```
[session copy 6.log](https://github.com/JMRI/JMRI/files/6446644/session.copy.6.log)

dsand47 commented:
> Unfortunately still fails:

The substitution occurs after the svg failure, so the exception dump still occurs.  

Does the icon substitution occur?  I have never used throttle so I don't know what should occur.

dheap commented:
@Lionel642

What about using:
`jmri.util.SystemType.isMacOSX()`

dheap commented:
> @dheap
> What about using:
> `jmri.util.SystemType.isMacOSX()`

But this may fail for macOS 11...

dsand47 commented:
> > @dheap
> > What about using:
> > `jmri.util.SystemType.isMacOSX()`
> 
> But this may fail for macOS 11...

It should be ok.  Internally, macOS 11 is reported as 10.16.
```
PanelPro version 4.23.5ish+das+20210508T2310Z+Rae51ce0d7c starts under Java 1.8.0_282 on Mac OS X x86_64 v10.16 at Sat May 08 18:10:22 CDT 2021 [main]
```

dheap commented:
@dsand47

Yes, that'll work. `isMacOSX()` returns `true` i `(lowerCaseName.contains("os x")` and thats what `System.getProperty("os.name")` is returning

> > > @dheap
> > > What about using:
> > > `jmri.util.SystemType.isMacOSX()`
> > 
> > But this may fail for macOS 11...
> 
> It should be ok.  Internally, macOS 11 is reported as 10.16.
> ```
> PanelPro version 4.23.5ish+das+20210508T2310Z+Rae51ce0d7c starts under Java 1.8.0_282 on Mac OS X x86_64 v10.16 at Sat May 08 18:10:22 CDT 2021 [main]
> ```
>

dheap commented:
@Lionel642

Why don't you replace the try-catch blocks with:

if jmri.util.SystemType.isMacOSX() {
    // Use .png image.
} else {
    // Use .svg image
}

A lot less ugly than throwing and catching exceptions.

dheap commented:
(Corrections. Can't edit on my old iPad.)

@Lionel642

Why don't you replace the try-catch blocks with:
``
if (jmri.util.SystemType.isMacOSX()) {
// Use .png image.
} else {
// Use .svg image
}
``
A lot less ugly than throwing and catching exceptions.

dsand47 commented:
The exception is happening outside of JMRI in thumbnailator so the try/catch does not help.  Use the approach from @dheap to prevent the call to thumbnailator when the OS is macOS.

Lionel642 commented:
Got it, svg files mime type was not correctly recognized on OSX, and only there. Now using file extension.
While at it : 
 - I cleaned the extra useless log messages, now just displaying something at debug level
 - I left the failover to use PNG if SVG loading fails.
 - I updated battik and its lib to a more recent version.

dsand47 commented:
The errors no longer occur on macOS 11.3.1.  I can open the throttle and click on the buttons and the light icon.

I don't actually use JMRI throttles so I can't confirm that there are no other issues.

Lionel642 commented:
reviewed failed tests, not related to modified code and ran ok locally

geowar1 commented:
Why not just convert all svg's to png's and have one filetype/code for all?

Lionel642 commented:
SVG is a vectorial image format, which allows for smooth resizing of images. 
That's the idea of this code update, it allows to resize the JMRI Throttle windows and its component as large as the user wishes, allowing to use JMRI native throttle with a Bluetooth device while standing up close to the tracks and far away from screen.

dheap commented:
With the latest changes, I'm not seeing any exceptions or log errors when I open a throttle and turn lights on/off on macOS 10.12.6 & 0.10.5.

geowar1 commented:
>>SVG is a vectorial image format
Then why not use a more cross platform vectorial image format like BMP?
I see no advantage of extra files and more complex code.
